pkg/ebpf/c/tracee.bpf.c

```diff
@@ -1965,40 +2005,154 @@ static __always_inline int init_tailcall_program_data(program_data_t *p, void *c
+    }
+    if (equality != NULL) {
+        equal_in_scopes = equality->equal_in_scopes;
+        set_in_scopes = equality->set_in_scopes;
+    }
 
-    return filter_out;
+    return equal_in_scopes | (filter_out_scopes & ~set_in_scopes);
 }
 
-static __always_inline int equality_filter_matches(bool filter_out, void *filter_map, void *key)
+static __always_inline u64 equality_filter_matches(u64 filter_out_scopes,
+                                                   void *filter_map,
+                                                   void *key)
 {
-    u32 *equality = bpf_map_lookup_elem(filter_map, key);
-    if (equality != NULL)
-        return *equality;
+    // check compute_scopes() for initial info
+    //
+    // e.g.: cmdline: -t 2:comm=who -t 3:comm=ping -t 4:comm!=who
+    //
+    // filter_out_scopes = 0000 1000, since scope 4 has "not equal" for comm filter
+    // filter_map        = comm_filter
+    // key               = "who" | "ping"
+    //
+    // ---
+    //
+    // considering an event from "who" command
+    //
+    // equal_in_scopes   = 0000 0010, since scope 2 has "equal" for comm filter
+    // set_in_scopes     = 0000 1010, since scope 2 and 4 are set for comm filter
+    //
+    // return            = equal_in_scopes | (filter_out_scopes & ~set_in_scopes)
+    //                     0000 0010 |
+    //                     (0000 1000 & 1111 0101) -> 0000 0000
+    //
+    //                     0000 0010 |
+    //                     0000 0000
+    //                     ---------
+    //                     0000 0010 = (scope 2 matched)
+    //
+    // considering an event from "ping" command
+    //
+    // equal_in_scopes   = 0000 0100, since scope 3 has "equal" for comm filter
+    // set_in_scopes     = 0000 0100, since scope 3 is set for comm filter
+    //
+    // return            = equal_in_scopes | (filter_out_scopes & ~set_in_scopes)
+    //                     0000 0100 |
+    //                     (0000 1000 & 0000 0100) -> 0000 0000
+    //
+    //                     0000 0100 |
+    //                     0000 0000
+    //                     ---------
+    //                     0000 0100 = (scope 3 matched)
 
-    return filter_out;
+    u64 equal_in_scopes = 0;
+    u64 set_in_scopes = 0;
+    eq_t *equality = bpf_map_lookup_elem(filter_map, key);
+    if (equality != NULL) {
+        equal_in_scopes = equality->equal_in_scopes;
+        set_in_scopes = equality->set_in_scopes;
+    }
+
+    return equal_in_scopes | (filter_out_scopes & ~set_in_scopes);
 }
 
-static __always_inline int bool_filter_matches(bool filter_out, bool val)
+static __always_inline u64 bool_filter_matches(u64 filter_out_scopes, bool val)
 {
-    return filter_out ^ val;
+    // check compute_scopes() for initial info
+    //
+    // e.g.: cmdline: -t 5:container
+    //
+    // considering an event from a container
+    //
+    //   filter_out_scopes = 0000 0000
```

Comment:
Based on tests, yes.

`-t 5:container`

config_entry:

```
"cont_filter_enabled_scopes": 16,
"cont_filter_out_scopes": 0,
```

val = **true** (from a container)
```c
filter_out_scopes ^ (val ? ~0ULL : 0)
 0 ^ -1 = -1
```

val = **false** (not from a container)
```c
filter_out_scopes ^ (val ? ~0ULL : 0)
 0 ^ 0 = 0
```

Reply:
`bool_filter_matches()` could return just `filter_out_scopes ^ -val;` without any prejudice.

Reply:
```
filter_out_scopes ^ (val ? ~0ULL : 0)
 0 ^ -1 = -1
```

> ... since otherwise all scopes will match, which is wrong

Isn't it indeed what we see here?